### PR TITLE
nextcloud26: init at 26.0.0

### PR DESCRIPTION
**Pull request by RaitoBezarius**
###### Description of changes

:tada: 
See this slide for relevant changelog: https://youtu.be/docYiLbGSxI?t=5058
https://nextcloud.com/blog/hub-4-pioneers-ethical-ai-integration-for-a-more-productive-and-collaborative-future/

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

closes https://github.com/NixOS/nixpkgs/pull/219929

### CODE REVIEW

File: pkgs/servers/nextcloud/patches/v26/0001-Setup-remove-custom-dbuser-creation-behavior.patch [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+diff --git a/lib/private/Setup/MySQL.php b/lib/private/Setup/MySQL.php
+index 50f566728a9..ce721415e30 100644
+--- a/lib/private/Setup/MySQL.php
++++ b/lib/private/Setup/MySQL.php
+@@ -145,62 +145,9 @@ class MySQL extends AbstractDatabase {
+ 	 * @param IDBConnection $connection
+ 	 */
+ 	private function createSpecificUser($username, $connection): void {
+-		$rootUser = $this->dbUser;
+-		$rootPassword = $this->dbPassword;
```
Comment (SuperSandro2000) [inline pkgs/servers/nextcloud/patches/v26/0001-Setup-remove-custom-dbuser-creation-behavior.patch:43]: so we can drop https://github.com/NixOS/nixpkgs/blob/nextcloud26/pkgs/servers/nextcloud/patches/v25/0001-Setup-remove-custom-dbuser-creation-behavior.patch#L37-L38 , too?

Comment (RaitoBezarius) [reply pkgs/servers/nextcloud/patches/v26/0001-Setup-remove-custom-dbuser-creation-behavior.patch:43]: as a cleanup PR for v25 yes, but I wouldn't bother too much? you can do it :P

**Comment by icewind1991** [comment]
the "remove custom dbuser creation" patch isn't required anymore for 26, you can now disable that behavior by setting `NC_setup_create_db_user=false` during setup or add `"setup_create_db_user" => false` to config.php before setup.

**Comment by RaitoBezarius** [comment]
> the "remove custom dbuser creation" patch isn't required anymore for 26, you can now disable that behavior by setting `NC_setup_create_db_user=false` during setup or add `"setup_create_db_user" => false` to config.php before setup.

Done.

File: nixos/modules/services/web-apps/nextcloud.nix [hunk 2]
```diff
@@ -747,7 +748,8 @@ in {
               `pkgs.nextcloud`.
             ''
           else if versionOlder stateVersion "22.11" then nextcloud24
-          else nextcloud25
+          else if versionOlder stateVersion "23.05" then nextcloud25
```
Comment (RaitoBezarius) [inline nixos/modules/services/web-apps/nextcloud.nix:751]: I feel like the logic is fine.

**Comment by RaitoBezarius** [comment]
@Ma27 do you think we should remove NC24?

File: nixos/tests/nextcloud/openssl-sse.nix [hunk 3]
```diff
@@ -71,12 +72,15 @@ in {
...
-        nextcloudwithopenssl1.fail("${withRcloneEnv3} ${pkgs.rclone}/bin/rclone cat nextcloud:test-shared-file >&2")
+        # This will succed starting NC26 because of their custom implementation of openssl_seal
+        read_existing_file_test = nextcloudwithopenssl1.fail if nextcloud_version < 26 else nextcloudwithopenssl1.succeed
+        read_existing_file_test("${withRcloneEnv3} ${pkgs.rclone}/bin/rclone cat nextcloud:test-shared-file >&2")
         nextcloudwithopenssl1.succeed("nextcloud-occ encryption:disable")
         nextcloudwithopenssl1.succeed("${copySharedFile3}")
         nextcloudwithopenssl1.succeed("grep bye /var/lib/nextcloud/data/root/files/test-shared-file2")
         nextcloudwithopenssl1.succeed("${withRcloneEnv3} ${pkgs.rclone}/bin/rclone cat nextcloud:test-shared-file2 | grep bye")
 
+
```
Comment (dotlambda) [inline nixos/tests/nextcloud/openssl-sse.nix:83]: Why the random newline?

Comment (RaitoBezarius) [reply nixos/tests/nextcloud/openssl-sse.nix:83]: sorri

**Comment by GaetanLepage** [comment]
Shouldn't it be added here as well ?
https://github.com/NixOS/nixpkgs/blob/60c1d71f2ba4c80178ec84523c2ca0801522e0a6/nixos/modules/services/web-apps/nextcloud.nix#L207

**Comment by Ma27** [comment]
I'll try to take a look later today, but no guarantees because exam phase.

File: nixos/modules/services/web-apps/nextcloud.nix (hunk 2 above)
Comment (Ma27) [reply nixos/modules/services/web-apps/nextcloud.nix:751]: Agreed.

**Comment by Ma27** [comment]
> @Ma27 do you think we should remove NC24?

Let's await the final patch release. Reason is quite simple: we can automatically backport all updates on the next updates and we don't have to fix up the backport.

Btw, I'd resolve the review comments and drop the nextcloud23 references (this makes the module simpler on some occasions) and deploy the changes to my setup. Would it be OK if I push the changes after that to your branch? :)

**Comment by RaitoBezarius** [comment]
> > @Ma27 do you think we should remove NC24?
> 
> Let's await the final patch release. Reason is quite simple: we can automatically backport all updates on the next updates and we don't have to fix up the backport.

Got it.

> Btw, I'd resolve the review comments and drop the nextcloud23 references (this makes the module simpler on some occasions) and deploy the changes to my setup. Would it be OK if I push the changes after that to your branch? :)

Please do so! It's running on my setup since the opening of this PR, no trouble so far.

**Comment by Ma27** [comment]
FWIW I got bitten by https://github.com/nextcloud/user_saml/issues/709, but considering that we don't ship SAML by default and the workaround is trivial I don't consider this a blocker.

**Review by Ma27 — APPROVED** [review]
LGTM. @RaitoBezarius would merge soonish if that's OK :)

**Comment by RaitoBezarius** [comment]
> @RaitoBezarius btw, why does [881efdb](https://github.com/NixOS/nixpkgs/commit/881efdb55e1480f0c0fef905874ce1d7c1c118a3) state that the stateVersion logic is broken? The change seems OK to me :)

Fixed, excessive amount of carefulness. :)

**Pull request merged**